src/index/TextIndexReadWrite.h

```diff
@@ -0,0 +1,190 @@
+/// vector which was of type T which lead to complicated handling of Ids.
+/// To see this look at the deleted lines 867, 888-896 of IndexImpl.Text.cpp
+/// Also an ongoing confusion is the naming of some variables since there
+/// has to happen two decodings thus we first have on variable each that can
+/// be named encoded.
+/// I tried to also template this function in a way to surpass the inconsistency
+/// between the datatypes written to the file and the datatypes we want to get
+/// out. This problem also stems from the fact that Posting is somewhat
+/// ambiguous since the WordIndex in a posting can refer to an entityId so a
+/// VocabIndex or a word so a WordVocabIndex.
+// Read a freqComprList from the textIndexFile. The From specifies the type
+// that was used to create the codebook in the writing step and the To
+// specifies the type to cast that codebook values to. This is done with a
+// static cast if no lambda function to cast is given.
+template <typename To, typename From>
+vector<To> readFreqComprList(
+    size_t nofElements, off_t from, size_t nofBytes,
+    const ad_utility::File& textIndexFile,
+    const std::function<To(From)>& transformer = [](From x) {
+      return static_cast<To>(x);
+    }) {
+  AD_CONTRACT_CHECK(nofBytes > 0);
+  LOG(DEBUG) << "Reading frequency-encoded list from disk...\n";
+  LOG(TRACE) << "NofElements: " << nofElements << ", from: " << from
+             << ", nofBytes: " << nofBytes << '\n';
+  size_t nofCodebookBytes;
+  vector<uint64_t> frequencyEncodedResult;
+  frequencyEncodedResult.resize(nofElements + 250);
+  off_t current = from;
+  size_t ret = textIndexFile.read(&nofCodebookBytes, sizeof(size_t), current);
+  LOG(TRACE) << "Nof Codebook Bytes: " << nofCodebookBytes << '\n';
+  AD_CONTRACT_CHECK(sizeof(size_t) == ret);
+  current += ret;
+  std::vector<From> codebook;
+  codebook.resize(nofCodebookBytes / sizeof(From));
+  ret = textIndexFile.read(codebook.data(), nofCodebookBytes, current);
+  current += ret;
+  AD_CONTRACT_CHECK(ret == size_t(nofCodebookBytes));
+  std::vector<uint64_t> simple8bEncoded;
+  simple8bEncoded.resize(nofElements);
+  ret = textIndexFile.read(simple8bEncoded.data(), nofBytes - (current - from),
+                           current);
+  current += ret;
+  AD_CONTRACT_CHECK(size_t(current - from) == nofBytes);
+  LOG(DEBUG) << "Decoding Simple8b code...\n";
+  ad_utility::Simple8bCode::decode(simple8bEncoded.data(), nofElements,
+                                   frequencyEncodedResult.data());
+  LOG(DEBUG) << "Reverting frequency encoded items to actual IDs...\n";
+  frequencyEncodedResult.resize(nofElements);
+  vector<To> result;
+  result.reserve(frequencyEncodedResult.size());
+  ql::ranges::for_each(frequencyEncodedResult, [&](const auto& encoded) {
+    result.push_back(transformer(codebook.at(encoded)));
+  });
+  LOG(DEBUG) << "Done reading frequency-encoded list. Size: " << result.size()
```

Comment:
In one of your next PRs you can refactor the reading and writing here, using the functionalities in
`util/serialization` , because all those manual `read` and `offset += ret` calls are very hard to read.